### PR TITLE
PYTHON-4256 OIDC Spec Cleanup

### PR DESCRIPTION
**Pull request by blink1073**
Requires https://github.com/mongodb-labs/drivers-evergreen-tools/pull/428

### CODE REVIEW

File: pymongo/auth_oidc.py [hunk 1]
```diff
@@ -189,30 +189,43 @@ def get_spec_auth_cmd(self) -> Optional[MutableMapping[str, Any]]:
...
+        # If authentication fails with error code 18, invalidate the access and
+        # refresh tokens, and try to authenticate again. If authentication fails for
+        # any other reason, raise the error to the user.
         if self.refresh_token:
             try:
                 return self._sasl_start_jwt(conn)
-            except Exception:  # noqa: S110
-                pass
+            except OperationFailure as e:
+                if e.code == 18:
```
Comment (NoahStapp) [inline pymongo/auth_oidc.py:225]: Could these three mostly-duplicate code blocks be abstracted into a helper method that takes an appropriate callback to handle the error code 18 case?

File: pymongo/helpers.py [hunk 2]
```diff
@@ -90,6 +90,9 @@
 # Server code raised when re-authentication is required
 _REAUTHENTICATION_REQUIRED_CODE: int = 391
 
+# Server code raised when authentication fails.
+_AUTHENTICATION_FAILURE_CODE: int = 18
```
Comment (NoahStapp) [inline pymongo/helpers.py:94]: Use this constant instead of the magic number `18` in `auth_oidc`?

File: test/auth_oidc/test_auth_oidc.py [hunk 3]
```diff
@@ -53,7 +53,7 @@
 TOKEN_FILE = os.environ.get("OIDC_TOKEN_FILE", "")
 
 # Generate unified tests.
-globals().update(generate_test_classes(str(TEST_PATH), module=__name__))
+# globals().update(generate_test_classes(str(TEST_PATH), module=__name__))
```
Comment (NoahStapp) [inline test/auth_oidc/test_auth_oidc.py:56]: Intended comment-out?

File: test/auth_oidc/test_auth_oidc.py [hunk 4]
```diff
@@ -107,15 +107,23 @@ def setUpClass(cls):
...
+    def setUp(self):
+        self.refresh_present = 0
+        super().setUp()
+
     def create_request_cb(self, username="test_user1", sleep=0):
         def request_token(context):
             # Validate the info.
             self.assertIsInstance(context.idp_info.issuer, str)
-            self.assertIsInstance(context.idp_info.clientId, str)
+            # self.assertIsInstance(context.idp_info.clientId, str)
```
Comment (NoahStapp) [inline test/auth_oidc/test_auth_oidc.py:118]: Intended comment-out?

File: test/auth_oidc/test_auth_oidc.py [hunk 5]
```diff
@@ -219,6 +227,26 @@ def test_1_6_allowed_hosts_blocked(self):
         # Close the client.
         client.close()
 
+    def test_1_7_allowed_hosts_in_connection_string_ignored(self):
+        # Create an OIDC configured client with the connection string: `mongodb+srv://example.com/?authMechanism=MONGODB-OIDC&authMechanismProperties=ALLOWED_HOSTS:%5B%22example.com%22%5D` and a Human Callback.
+        # Assert that the creation of the client raises a configuration error.
+        uri = "mongodb+srv://example.com?authMechanism=MONGODB-OIDC&authMechanismProperties=ALLOWED_HOSTS:%5B%22example.com%22%5D"
+        with self.assertRaises(ConfigurationError), warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            _ = MongoClient(
```
Comment (NoahStapp) [inline test/auth_oidc/test_auth_oidc.py:236]: Pointless assignment just for the linter?

**Review by NoahStapp — CHANGES_REQUESTED** [review]
A few clarifying questions, otherwise looks good!

File: pymongo/helpers.py (hunk 2 above)
Comment (blink1073) [reply pymongo/helpers.py:94]: Done

File: test/auth_oidc/test_auth_oidc.py (hunk 3 above)
Comment (blink1073) [reply test/auth_oidc/test_auth_oidc.py:56]: Fixed

File: test/auth_oidc/test_auth_oidc.py (hunk 5 above)
Comment (blink1073) [reply test/auth_oidc/test_auth_oidc.py:236]: Yes

**Review by NoahStapp — APPROVED** [review]
I like the validator approach a lot!

**Pull request merged**